Utils/Subject.h

```diff
@@ -1,104 +1,270 @@
 	};
 
 	template <class T>
 	class TSubject
 	{
-	public:
-		TSubject() = default;
-		~TSubject()
+	private:
+		struct MemberFunctionWrapper
 		{
-			for (auto& observer : m_Observers)
+			virtual ~MemberFunctionWrapper() = default;
+			virtual void Call(const T& value) const = 0;
+			virtual unsigned int GetId() const = 0;
+			virtual bool IsConst() const = 0;
+		};
+
+		template<typename TClass>
+		struct ImplMemberFunctionWrapper final : public MemberFunctionWrapper
+		{
+			ImplMemberFunctionWrapper(TClass* _pThis, void(TClass::* _pFunction)(const T&))
+				: pThis{ _pThis }, pFunction{ _pFunction }, id{ ReflectionUtils::GenerateTypenameHash<TClass>() }
 			{
-				observer->OnSubjectDestroy();
 			}
-		}
+			virtual ~ImplMemberFunctionWrapper() = default;
+
+			unsigned int id{};
+			TClass* pThis{};
+			void(TClass::* pFunction)(const T&) {};
+
+			virtual void Call(const T& value) const override
+			{
+				(pThis->*pFunction)(value);
+			}
+			virtual unsigned int GetId() const override { return id; }
+			virtual bool IsConst() const override { return false; }
+		};
+
+		template<typename TClass>
+		struct ConstImplMemberFunctionWrapper final : public MemberFunctionWrapper
+		{
+			ConstImplMemberFunctionWrapper(TClass* _pThis, void(TClass::* _pFunction)(const T&) const)
+				: pThis{ _pThis }, pFunction{ _pFunction }, id{ ReflectionUtils::GenerateTypenameHash<TClass>() }
+			{
+			}
+			virtual ~ConstImplMemberFunctionWrapper() = default;
+
+			unsigned int id{};
+			TClass* pThis{};
+			void(TClass::* pFunction)(const T&) const {};
+
+			virtual void Call(const T& value) const override
+			{
+				(pThis->*pFunction)(value);
+			}
+			virtual unsigned int GetId() const override { return id; }
+			virtual bool IsConst() const override { return true; }
+		};
 
-		void AddListener(TObserver<T>* observer)
+	public:
+		TSubject() = default;
+
+		template<typename TClass>
+		void AddListener(TClass* pThis, void(TClass::* pFunction)(const T&))
 		{
-			m_Observers.push_back(observer);
+			std::unique_ptr<MemberFunctionWrapper> pFunctionWithThis{ std::make_unique<ImplMemberFunctionWrapper<TClass>>(pThis, pFunction) };
+			m_pMemberFunctions.emplace_back(std::move(pFunctionWithThis));
+		}
+		template<typename TClass>
+		void AddListener(TClass* pThis, void(TClass::* pFunction)(const T&) const)
+		{
+			std::unique_ptr<MemberFunctionWrapper> pFunctionWithThis{ std::make_unique<ConstImplMemberFunctionWrapper<TClass>>(pThis, pFunction) };
+			m_pMemberFunctions.emplace_back(std::move(pFunctionWithThis));
 		}
 		void AddListener(void(*pCallback)(const T&))
 		{
 			m_Functions.push_back(pCallback);
 		}
-		void RemoveListener(TObserver<T>* observer)
+		template<typename TClass>
+		void RemoveListener(TClass* pThis, void(TClass::* pFunction)(const T&))
+		{
+			constexpr unsigned int classId{ ReflectionUtils::GenerateTypenameHash<TClass>() };
+			for (int i{ static_cast<int>(m_pMemberFunctions.size()) - 1 }; i >= 0; --i)
```

Comment:
Cache vector size. Every for tick the size will get recalculated.

Reply:
Vector size is only used for initializing the for index